samcli/lib/pipeline/bootstrap/stage.py

```diff
@@ -0,0 +1,280 @@
+        self.ecr_repo: EcrRepo = EcrRepo(arn=ecr_repo_arn)
+
+    def did_user_provide_all_required_resources(self) -> bool:
+        """Check if the user provided all of the stage resources or not"""
+        return (
+            self.pipeline_user.is_user_provided
+            and self.pipeline_execution_role.is_user_provided
+            and self.cloudformation_execution_role.is_user_provided
+            and self.artifacts_bucket.is_user_provided
+            and (not self.create_ecr_repo or self.ecr_repo.is_user_provided)
+        )
+
+    def _get_non_user_provided_resources_msg(self) -> str:
+        missing_resources_msg = ""
+        if not self.pipeline_user.is_user_provided:
+            missing_resources_msg += "\n\tPipeline user"
+        if not self.pipeline_execution_role.is_user_provided:
+            missing_resources_msg += "\n\tPipeline execution role."
+        if not self.cloudformation_execution_role.is_user_provided:
+            missing_resources_msg += "\n\tCloudFormation execution role."
+        if not self.artifacts_bucket.is_user_provided:
+            missing_resources_msg += "\n\tArtifacts bucket."
+        if self.create_ecr_repo and not self.ecr_repo.is_user_provided:
+            missing_resources_msg += "\n\tECR repo."
+        return missing_resources_msg
+
+    def bootstrap(self, confirm_changeset: bool = True) -> bool:
+        """
+        Deploys the CFN template(./stage_resources.yaml) which deploys:
+            * Pipeline IAM User
+            * Pipeline execution IAM role
+            * CloudFormation execution IAM role
+            * Artifacts' S3 Bucket along with KMS encryption key
+            * ECR Repo
+        to the AWS account associated with the given stage. It will not redeploy the stack if already exists.
+        This CFN template accepts the ARNs of the resources as parameters and will not create a resource if already
+        provided, this way we can conditionally create a resource only if the user didn't provide it
+
+        THIS METHOD UPDATES THE STATE OF THE CALLING INSTANCE(self) IT WILL SET THE VALUES OF THE RESOURCES ATTRIBUTES
+
+        Parameters
+        ----------
+        confirm_changeset: bool
+            if set to false, the stage_resources.yaml CFN template will directly be deployed, otherwise, the user will
+            be prompted for confirmation
+
+        Returns True if bootstrapped, otherwise False
+        """
+
+        if self.did_user_provide_all_required_resources():
+            click.secho(f"\nAll required resources for the {self.name} stage exist, skipping creation.", fg="yellow")
+            return True
+
+        missing_resources_msg: str = self._get_non_user_provided_resources_msg()
+        click.echo(
+            "This will create the following required resources for the {self.name} stage: " f"{missing_resources_msg}"
+        )
+        if confirm_changeset:
+            confirmed: bool = click.confirm("Should we proceed with the creation?")
```

Comment:
Is this option related to CloudFormation ChangeSets? (as is the case for `sam deploy`)

Reply:
yes, used like here https://github.com/aws/aws-sam-cli/blob/d965ee4d3228c8f7fd76e2cb083f22a0d4841ec7/samcli/commands/deploy/deploy_context.py#L237